Feature/block bootstrap

paulf81 opened the pull request:
is NOT ready to be merged

**Feature or improvement description**
This is a first draft implementation of block bootstrapping.  In our current bootstrapping, we sample all R rows from a dataframe with R totals rows (with replacement) N times.  In the proposed block bootstrapping method, the data are divided into D consecutive blocks (with the 0th block including any residual rows at the end), and we sample D of these blocks (with replacement) N times instead.   I think theory is roughly like:

https://en.wikipedia.org/wiki/Bootstrapping_(statistics)#Block_bootstrap

But essentially since we have 1/10 minute samples, these might relate a lot in time, so removing whole blocks from iterations could help to better show underlying uncertainty

I added a demonstration to the example: energy_ratio_vs_wd_for_single_df, but should work also with suite

**Related issue, if one exists**
Issue #39

paulf81 commented:
@misi9170 same question to you

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio.py:
```diff
@@ -299,6 +300,8 @@ def get_energy_ratio(
         N=1,
         percentiles=[5.0, 95.0],
         return_detailed_output=False,
+        block_bootstrapping = False, 
```
Remove spaces around equal sign for consistency

paulf81 replied:
Done!

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio.py:
```diff
@@ -405,6 +408,8 @@ def get_energy_ratio(
             N=N,
             percentiles=percentiles,
             return_detailed_output=return_detailed_output,
+            block_bootstrapping = block_bootstrapping, 
```
Same here

paulf81 replied:
Done!

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio.py:
```diff
@@ -504,6 +509,8 @@ def _get_energy_ratios_all_wd_bins_bootstrapping(
     N=1,
     percentiles=[5.0, 95.0],
     return_detailed_output=False,
+    block_bootstrapping = False, 
```
and here

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio.py:
```diff
@@ -504,6 +509,8 @@ def _get_energy_ratios_all_wd_bins_bootstrapping(
     percentiles=[5.0, 95.0],
     return_detailed_output=False,
+    block_bootstrapping = False, 
+    num_blocks = 10
```
I wonder if somehow we can make this more uniform. Essentially, regular bootstrapping is block bootstrapping with num_blocks==num_samples, correct? Maybe that can simplify things like that and that comparison can also work as a unit test.

Then, we can collapse the two options `block_bootstrapping` and `num_blocks` just to `num_bootstrapping_blocks`, with `-1` being the default option meaning `num_blocks=num_samples`? Or maybe that would all become too ambiguous. What do you think?

paulf81 replied:
I like it in principle, my only doubt is if the block-bootstrapped version of normal bootstrapping with num_blocks=num_samples is slower than the present implementation?  I should test, if that's ok, I think it makes sense.  What do you think @misi9170 ?

Bartdoekemeijer replied:
I agree, would be good to make sure things don't slow down!

paulf81 replied:
hi @Bartdoekemeijer  I tried this by putting in this code block around line 670 in energy_ratio.py:

```
        if not block_bootstrapping:
            num_blocks = df_binned.shape[0]
            block_bootstrapping = True
```
Unfortunately it was about 10x slower, I think maybe because the other call functions through built-in pandas (numpy) methods it gets a big speed up.   So maybe we keep the logic as is? 

 Still though, I think we could have the function inputs act as you like, instead of the seperate boolean, just haave num_blocks default to -1, meaning no blocks, does that work for you?

Bartdoekemeijer replied:
Yes, that works! Thanks for checking, and I agree that a 10x slowdown is not worth it.

misi9170 replied:
I'd say feel free to merge if this does not affect current code functionality. Given the slowdown, it's best to keep the bootstrapping and block bootstrapping separate.

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio.py:
```diff
@@ -655,11 +666,33 @@ def _get_energy_ratio_single_wd_bin_bootstrapping(
+
+        # If using block-bootstrapping, set up blocks
+        if block_bootstrapping:
+            block_list = list(range(num_blocks))  # List of all block names
```
I haven't tried this, but I wonder how things work if you have 10 samples and split it over 3 blocks. It should split it into 3 vectors with length 4, 3 and 3, respectively. And if you have 8 samples and split it into 3 blocks, it should have three vectors with length 3, 3 and 2. Do those work out, currently? My thought with the current line of code ` block_length = int(len(block_indices) / num_blocks)` was that this may not be resolved properly.

Maybe consider `np.split_array` is a good function for this. For example
```
sample_indices = np.arange(0, num_samples, dtype=int)
block_indices = np.array_split(sample_indices, num_blocks)
```

paulf81 replied:
I like that, I'll give it a shot!

paulf81 replied:
hi @Bartdoekemeijer  I remember what I did here, I start out by assigning everything to the first (0th block) with this line:

```
block_labels = np.zeros(len(block_indices)).astype(int) # Labels to assign each index
```

then I assign the blocks here:

```
for b_i in range(num_blocks):
    block_labels[block_length * b_i:block_length * (b_i+1)] = b_i
```

If the things don't divide evenly the leftovers get assigned to the first (0th block) which I guess could make that block a little bigger than the others but it gives some wrapping around, which I thought could be desirable.  But maybe your method is still preferable since it makes sure things are more even

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio_suite.py:
```diff
@@ -315,6 +315,8 @@ def get_energy_ratios(
         percentiles=[5.0, 95.0],
         balance_bins_between_dfs=True,
         return_detailed_output=False,
+        block_bootstrapping = False, 
```
remove spaces adjacent to equal sign

paulf81 replied:
Done!

Bartdoekemeijer commented on flasc/energy_ratio/energy_ratio.py:
```diff
@@ -655,11 +666,33 @@ def _get_energy_ratio_single_wd_bin_bootstrapping(
+            if not block_bootstrapping:
+                df_randomized = df_binned.sample(frac=1, replace=True).copy()
+            else:
+                test_blocks = choices(block_list, k = num_blocks) # Choose a set of blocks
```
I like this and I think this can also work for regular bootstrapping if num_blocks == num_samples.

Bartdoekemeijer reviewed (changes requested):
Thanks for doing this, Paul! See my comments.

paulf81 commented:
hi @Bartdoekemeijer and @misi9170 I think I've updated now as we discussed, block-bootstrapping is initiated via setting  num_blocks something other than the default of -1.  
I updated the example and also updated the docstrings.
And I added a valueError check on num_blocks
@misi9170  what do you think, good to merge?
